### DISCUSSION TITLE
"Stacking" extras without invoking `pip` conflict resolution

### DISCUSSION THREAD

**Discussion by silv-io** [pypa/setuptools#4279] [Q&A] [answered]
We have recently switched from using `setup.cfg` to `pyproject.toml`. In our `setup.cfg`, we had "stacked" extras like this:

```
base-runtime:
    package

runtime:
    %(base-runtime)s
    other-package
```

So, installing the extra `runtime` meant that all the packages in `base-runtime` are also installed. 

Now, with the `pyproject.toml`, to do this I've only found https://github.com/pypa/setuptools/discussions/3627, where it is suggested to depend on `project[base-runtime]` in the `runtime` extra.

However, what this does is, instead of just inserting all the dependencies of `base-runtime` in `runtime`, it invokes an extra conflict resolution step via `pip`.

With a sufficiently big enough set of dependencies and across projects which depend on each other, this makes using tools like `pip-tools` to create dependency pins a very slow process. (Tenfold increase from 4.5 minutes to 45 minutes).

So, is there a way to make the stacking of extras work like with the `setup.cfg` (presumably just inserting the strings into the array of the extra)?

**Comment by abravalheri** [comment, marked as answer]
Hi @silv-io, the `%(base-runtime)s` syntax is a feature of Python's `configparser` module. There is no equivalent for TOML (the TOML syntax does not include interpolation features).

So I am afraid that there is no way to achieve that, other than `project[base-runtime]` (which depends on `pip`).
